air/src/stack/op_flags/mod.rs

```diff
@@ -501,101 +920,147 @@ impl<E: FieldElement> EvaluationFrameExt<E> for &EvaluationFrame<E> {
-        .take(8)
-        .for_each(|v| *v *= not_3);
-    degree_7_intermediate_values
-        .iter_mut()
-        .take(16)
-        .skip(8)
-        .for_each(|v| *v *= frame.op_bit(3));
-    degree_7_intermediate_values.copy_within(0..16, 16);
-
-    let not_4 = binary_not(frame.op_bit(4));
-    degree_7_intermediate_values
-        .iter_mut()
-        .take(16)
-        .for_each(|v| *v *= not_4);
-    degree_7_intermediate_values
-        .iter_mut()
-        .take(32)
-        .skip(16)
-        .for_each(|v| *v *= frame.op_bit(4));
-    degree_7_intermediate_values.copy_within(0..32, 32);
-
-    let not_5 = binary_not(frame.op_bit(5));
-    let not_6 = binary_not(frame.op_bit(6));
-
-    // The first op_bit for a degree seven operation is 0. In order to save redundant
-    // multiplication ops and given the first bit will always be 0, a binary_not is
-    // being multiplied with both 2nd op bit as well as it's boolean not.
-    let not5_not6 = not_5 * not_6;
-    let not6_5 = not_6 * frame.op_bit(5);
-    degree_7_intermediate_values
-        .iter_mut()
-        .take(32)
-        .for_each(|v| *v *= not5_not6);
-    degree_7_intermediate_values
-        .iter_mut()
-        .take(64)
-        .skip(32)
-        .for_each(|v| *v *= not6_5);
-
-    degree_7_intermediate_values
-}
-
-/// This helper function generates flag values of all the degree six and four stack
-/// operations.
-fn set_op_flags_degree_4_and_6<E: FieldElement>(
+/// This helper function generates the flag values of all the individual and composite stack
+/// operations. This includes the following operations:
+/// - degree 7 operations which doesn't shift the stack.
+/// - degree 7 operations which shifts the stack to the left.
+/// - degree 7 operations which shifts the stack to the right.
+/// - degree 6 and 4 operations.
+/// - composite flags for individual stack items whose value has been copied over.
+/// - composite flags for individual stack items which has been shifted to the left.
+/// - composite flags for individual stack items which has been shifted to the right.
+/// - composite flag for the stack if the stack has been shifted to the left.
+/// - composite flag for the stack if the stack has been shifted to the right.
+/// - composite flag if the current operation being executed is a control flow operation or not.
+fn generate_op_flags<E: FieldElement>(
     frame: &EvaluationFrame<E>,
-) -> ([E; NUM_DEGREE_6_OPS], [E; NUM_DEGREE_4_OPS]) {
+) -> (
+    [E; NUM_DEGREE_7_OPS],
+    [E; NUM_DEGREE_6_OPS],
+    [E; NUM_DEGREE_4_OPS],
+    [E; NUM_NO_SHIFT_FLAGS],
+    [E; NUM_NO_SHIFT_FLAGS],
+    [E; NUM_RIGHT_SHIFT_FLAGS],
+    E,
+    E,
+    E,
+) {
+    // intermediary array to cache the value of intermediate flags.
+    let mut degree_7_intermediate_values = [E::ZERO; NUM_DEGREE_7_OPS];
     let mut degree_6_intermediate_values = [E::ZERO; NUM_DEGREE_6_OPS];
     let mut degree_4_intermediate_values = [E::ZERO; NUM_DEGREE_4_OPS];
```

Comment:
nit: maybe we should change the names of these to something like `degree7_op_flags`, `degree6_op_flags` etc. Would make it much less verbose and I don't think it would confuse things too much.